### ISSUE TITLE
Package `primitives`

### ISSUE THREAD

**Issue by pevab** [calicarpa/krum#27]
Implement the package `primitives` (cf adr below, also in the repo).

definition of done
- [ ] primitives are implemented
- [ ] unit tested
- [ ] reference doc updated

---

# ADR 2026-05-22

The library is organized in (for now) two packages
- primitives
- simulations
- orchestration

# Package `primitives`

The `primitives` package contains
- Aggregator interface and implementations
- Attack interface and implementations
- Model class (encapsulating a nn.Module)
    - parameters: Tensor
    - gradients: Tensor

These classes (especially the class Model) are designed so that
- the end-user should never care about flattening tensors
- the end-user will have to care about tensor copy, swap and sharing (parameters, or gradients)

# Package `simulations`

The `simulations` package contains reusable simulations of well-known papers.

Single parameter server
- [Machine learning with adversaries: byzantine tolerant gradient descent](https://papers.nips.cc/paper_files/paper/2017/file/f4b9ec30ad9f68f89b29639786cb62ef-Paper.pdf)
- [The hidden vulnerability](https://infoscience.epfl.ch/server/api/core/bitstreams/1efca80d-0b20-47ef-807f-c3cb744a8caa/content)

Decentralized
- Monna [Robust collaborative learning with linear gradient overhead](https://arxiv.org/pdf/2209.10931)

# Package `orchestration`

The `orchestration` package contains the logic to manage 1000s of simulation runs
and process their results. Will be designed in a later discussion.

**Comment by moh** [comment]
Question about the `Attack` base class API.

For aggregators, the current API stores `n` and `f` in the constructor:

```python
aggregator = Krum(n=n, f=f)
aggregated = aggregator(gradients)
```

This makes sense because `n` and `f` are part of many GAR definitions and validity constraints. For example, Krum, MultiKrum, and Bulyan use `f` directly in their selection rules, worker-count requirements, theoretical bounds, and sometimes complexity.

For attacks, I see two possible APIs:

```python
attack = NaNAttack()
byzantine_gradients = attack(honest_gradients, num_byzantine=f)
```

or:

```python
attack = NaNAttack(f=f)
byzantine_gradients = attack(honest_gradients)
```

The first option treats the attack object as a reusable strategy. This fits simple omniscient attacks such as NaN and Inf, where `f` only controls how many Byzantine gradients are generated and does not change the attack logic.

The second option is more consistent with the aggregator constructors and makes simulations slightly cleaner when `f` is fixed for the whole run.

My current preference is the first option:

```python
attack(honest_gradients, num_byzantine) -> byzantine_gradients
```

because it keeps simple attacks stateless and avoids storing a value that is not really part of the attack strategy.

Does this direction make sense, or should attacks also store `f` in the constructor for consistency with GARs?

**Comment by ArthurDanjou** [comment]
est-ce qu'on parle de `n`, `m` et `f` ou de `num_workers`, `num_byzantines`...
